[A8W8 GEMM] Optimized weight-preshuffled implementation & add quantization datatype for CK TILE rms_norm

aska-0096 opened the pull request:
## Proposed changes

Please describe the motivation behind the pull request, whether it enables a new feature or fixes a bug. If there are associated pull requests or issues, please link them to the pull request.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [x] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

aska-0096 committed:
tempsave

aska-0096 committed:
temp save

aska-0096 committed:
tempsave

aska-0096 committed:
tempsave, epilogue optimization for universal gemm done. TODO: mulitpleD epilogue optimization

aska-0096 committed:
temp save

aska-0096 committed:
tempsave

aska-0096 committed:
Merge branch 'transpose_opt' of https://github.com/ROCm/composable_kernel into rowwise_opt

aska-0096 committed:
temp save

aska-0096 committed:
update bf16 instance list

aska-0096 committed:
clang format

aska-0096 committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into mem_gemm_opt

aska-0096 committed:
bug fix

aska-0096 committed:
temp save

aska-0096 committed:
tempsave

aska-0096 committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into mem_gemm_opt

aska-0096 committed:
revert exp changes.

aska-0096 committed:
add blank line

junhaha666 committed:
add int8 gemm multiply multiply a8w8

junhaha666 committed:
uncomment

junhaha666 committed:
clang-format-12

junhaha666 committed:
Add example_gemm_multiply_multiply_xdl_int8

junhaha666 committed:
Remove shell scripts

aska-0096 committed:
Merge branch 'develop' into gemm_multiply_multiply_int8a8w8

aska-0096 committed:
update preprocess number for mi308; bring back printout in ckprofiler

aska-0096 committed:
tempsave

aska-0096 committed:
update ck_a8w8 library, update flush cache timing api

aska-0096 committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into update_cka8w8

aska-0096 committed:
remove the change in ckprofiler src

aska-0096 committed:
clean the flush_cache api

aska-0096 committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into update_cka8w8

aska-0096 committed:
reduce prefetch stage in blockwisepipev4

aska-0096 committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into update_cka8w8

aska-0096 committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into update_cka8w8

aska-0096 committed:
update tile size for fp8 rowwise

aska-0096 committed:
Merge branch 'mem_gemm_opt' of https://github.com/ROCm/composable_kernel into update_cka8w8

aska-0096 committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into update_cka8w8

aska-0096 committed:
fix bug in enable f8 gemm inside ckProfiler

aska-0096 committed:
update instance and lds layout strategy

aska-0096 committed:
delete use less files

aska-0096 committed:
fix cmake bug

aska-0096 committed:
update instances

aska-0096 committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into update_cka8w8

coderfeli committed:
add configs to fix tunning cases

coderfeli committed:
add configs to fix tunning cases

coderfeli committed:
port tiles from a8w8

coderfeli committed:
rm debug used files

coderfeli committed:
add instances

coderfeli committed:
remove all non gemm in cmake

coderfeli committed:
fix build

coderfeli committed:
fix build

aska-0096 committed:
sanity bug fix

coderfeli committed:
add bypass logic and build

coderfeli committed:
can run

coderfeli committed:
add double buffer scratch

aska-0096 committed:
remove agpr usage when vgpr usage <256

coderfeli committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into update_cka8w8

coderfeli committed:
fix performance regression on blockgemm v3 pipe

coderfeli committed:
using develop branch timer

coderfeli committed:
impl fp16 in ckprofiler

coderfeli committed:
Merge branch 'update_cka8w8' into update_cka8w8_uc

coderfeli committed:
add cpu shuffle

coderfeli committed:
fix tail

aska-0096 committed:
use empty hipstream in ckprofiler

coderfeli committed:
Merge remote-tracking branch 'origin/cka8w8_devtimer' into update_cka8w8_uc

coderfeli committed:
fix missed files and fix clang format

coderfeli committed:
fix fp16 build

coderfeli committed:
fix cmake rm compile options

coderfeli committed:
fix brepeat, kloop and lds two buffer; works ok now

coderfeli committed:
use new pipeline for b preshuffle, run ok; revert olds to fix ckprofiler

coderfeli committed:
auto calculate hard code params

coderfeli committed:
fix warnings and revert cmake and fix clang format

aska-0096 committed:
tempsave

aska-0096 committed:
Merge branch 'dev/a8w8_b_preshuffle' of https://github.com/ROCm/composable_kernel into add_a8w8_preshuffle_ckprofiler

aska-0096 committed:
sanity pass, most tile size enabled. TODO: NWave!=4

aska-0096 committed:
disable N, K  Padding, splitk enabled

aska-0096 committed:
add fp16 instances

aska-0096 committed:
use bpreshuffle as independent example

aska-0096 committed:
Merge branch 'add_a8w8_preshuffle_ckprofiler' of https://github.com/ROCm/composable_kernel into update_cka8w8_uc

aska-0096 committed:
refine weight preshuffle format.

aska-0096 committed:
tempsave

aska-0096 committed:
optimize software pipeline

aska-0096 committed:
refine blockgemm pipeline version as base struct.

aska-0096 committed:
fp8 add_rmsnorm_dynamic_dequant

aska-0096 committed:
add save_x=true instance

aska-0096 committed:
tempsave

aska-0096 committed:
Add compute-friendly pipeline for bpreshuffle case; remove enable-post-misched=0 flag.

aska-0096 committed:
fix Odd Mrepeat number pipelinev3; Add v3 instances to ckProfiler

aska-0096 committed:
clean the code

aska-0096 committed:
Merge pull request #1838 from ROCm/cka8w8_uc_newpipe

Cka8w8 uc newpipe

illsilin committed:
Merge from internal (#1857)

* enable batched_gemm_softmax_gemm_perm_wmma for gfx12

* disable instances with blocksize=256 in attention examples

* debuggging

* debug

* fixed lds_enabled

* debugging

* Fix and add limit to skiplds feature

* Enable skipLds feature and fix compilation bugs

* add ck_tile definitions for gfx12

* fix clang format and test/wmma_op

* updage instances cmake for gfx12

* disable the test_wmma_op on gfx12

* fix the builds for gfx950

* add gfx12 and gfx950 to default target list

* clean-up cmake file

* Initial introduction of OFP8 data types.

* Renamed FP8 and BF8 tests into FP8_FNUZ and BF8_FNUZ.

* Implementation of ConvertFP32Nearest in test_fp8_ocp.

* Remove dependence on possibly undeclared alias.

* Implement FP8OCP test for stochastic rounding mode.

* Implement FP8OCP tests for half_t type conversions.

* enable bf16 atomic add on gfx950

* Implement ConvertFP32Nearest test.

* Implement ConvertFP32Stochastic test.

* Implement ConvertFP16Nearest and ConvertFP16Stochastic tests.

* Refactoring. Move FP8 definitions into a separate header file.

* Enable easy switching between architectures.

* Fix compilation error for gfx942 architecture.

* Add fp4 type with constants

* only builf gfx950 branch for gfx950 target by default

* Enable OCP build of example_gemm_xdl_fp8.

* Fix formatting.

* fix the build logic for gfx950

* Improve GEMM example verbosity.

* Add constexpr where applicable.

* fix the logic of enabling XDL and WMMA instances

* Improve GEMM example verbosity.

* Enable build of example_gemm_xdl_fp8_bf8 test.

* Fix tests for gfx1101 architecture.

* Build DPP examples only on gfx103 and gfx11 architectures.

* Optionaly run either CPU or GPU verifications with GEMM examples.

* Extend GeneratorTensor_Sequential to produce values of prescribed data types.

* Add missing constructor.

* Add scale type and mxfp conversions

* Update conversions

* Add conversion tests

* Fix typo

* Improve infrastructure for OFP8 data type support.

* BUGFIX. Should not use FP8 as Compute/Accum data type.

* Add custom target for grouped_convnd_bwd_weight tests.

* Can build `tests` target on gfx950.

* Bugfixes on gfx1101 architecture.

* Fix dependencies.

* Add stochastic rounding tests

* Provide single point of truth for FP8 INF and NAN checks

* Prevent instantiation of operators that are not supported by FP8 data types

* Add FP8 type selection into client_axample CMakeLists.txt

* Prevent sccache server from shutting down during build

* Fix test success reporting logic

* Change default verification method to CPU.

GPU verification takes too much time to complete on the emulator.

* Add scale <-> float conversions

* Add scaled conversions with tests

* Add device conversions

* Make sure all tests and examples are built for gfx950

* Facilitate testing of FP8 data types on the emulator

* Introduce two new tensor generators

* Enable instances built for gfx94 to be built on gfx950

* Verify 35_splitk_gemm on floating point numbers.

splitk gemm appears to be losing precision VS reference implementation when FP numbers are involved.

* Format

* Verify 04_gemm_add_add_fastgelu on floating point numbers

* Verify 20_grouped_conv_bwd_weight on floating point numbers

* Verify 38_grouped_conv_bwd_data_multiple_d on floating point numbers

* Verify more tests on floating point data

* Fix data types and improve testing verbocity.

* Add fp4 vectors

* Add debug tests

* Upgrade to NPI 573 build docker.

* Skip on gemm_universal tests.

The tests take too long to complete on the emulator.
Need to see if it is possible to reduce the scope of the testing to just FP8 data types.

* Add new mfma instructions and examples

* Add preprocessor directives for gfx950 specific code

* Fix gfx1101 build

* Document test availability

* Re-enable fp8 gemms for gfx94/95

* Cherry-pick GEMM Universal tests for FP8 data types

* Cleanup

* Add vector types and tests

* Add check_err function

* Add tensor generators

* CK_USE_GFX94 has already been set on this branch

* Fix

* Address formatting issues and leftovers

* Make fail/pass logic consistent within 01_gemm folder

Removed multiple negations in fail/pass logic to propagate `true` as the success indicator.

* Fix GPU verification reporting logic.

* Update year in copyright notice.

* Cleanup

* Use `enum class` instead of `enum`

* Remove set_property for FP8 tests

* Add vector conversions

* Fix

* Fix linker errror

* Clean up

* Fix gfx950 conversions

* Clean up

* Fix more gfx950 conversions

* Fix even more gfx950 conversions

* Narrowing the scope of PR to OCP FP8 enablement only

* Add tests for OCP FP8 vector_type storage

* Fix client examples build

* Fix typo

* Update e8m0 casting

* Rename E8M0 type

* Update unpack method

* Cleanup merge artifacts

* Enable gemm kernel on all gfx9 architectures (#227)

* clean-up

* Implement `non_native_vector_base` with `ext_vector_type` array. (#232)

* Enable support of 1, 2, 4, and 8-byte custom types in CK.

* Fix pool tests for OCP FP8 data type

* Fix build

* Add ckProfiler gemm instances for new mfma instructions and fix ckProfiler build on MI350

* fix clang format

* Add new mfma instructions and examples

* Add preprocessor directives for gfx950 specific code

* Add ckProfiler gemm instances for new mfma instructions and fix ckProfiler build on MI350

* fix clang format

* Fix clang format for the newly merged files

* Use the existing example instances for fp16 bf16 and int8

* Remove comment on new mfma instructions in MfmaInstr

* Update include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* merge from public repo

* Fix ck build

* Fix ck build

* Use double for max_abs_in_val

* Move scaled_type_convert functions to a separate header (#251)

* re-enable building mha lib and gemm_universal_f8 instances for gfx950

* Update library/src/tensor_operation_instance/gpu/CMakeLists.txt

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* fix typo for CK_USE_OCP_FP8

* fix typo for CK_USE_OCP_FP8

* Add FP6 and BF6 types (#261)

* Add a rounding flag

* Add FP6 and BF6

* Add tests

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* Clean up

---------

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* fix one more typo

* Refactor E8M0 scale implementation (#262)

* Refactor E8M0 scale implementation

* Add MXFP6 and MXBF6 conversion methods (#270)

* Add conversions

* Add tests

* Add docstrings

* Add scaled conversions

* Add fp6/bf6 tests

* Remove misleading fp4 test case

* Add docstrings

* Clean up

* Address comments

* Set stricter tolerances for RNE tests

* Add missing tests

* Add native conversions to float

* Revert "Add native conversions to float"

This reverts commit 09467111f73b753c8cc3d597533b187940353dab.

* Update copyright years

* replace the fp6 with bf6 convert calls in test_bf6

* fix test_bf6

* enable smfmac test

* [MX FP8] Add Scaled Type Convert Functions for OCP FP8/BF8 data types (#271)

* Move scaled_type_convert functions to a separate header

* Introduce MX data tests

* Build MX tests only on relevant architectures

* Refactor E8M0 scale implementation

* Fix `config.h` typo

* Cleanup deprecated symbols

* Refactor `amd_ck_fp8.hpp`

* `scaled_type_convert` for `f8_ocp_t`

* Implement test for MX FP8 scaled type convert

* Implement test for MX BF8 scaled type convert

* Scaled type convert for vectors of 2 FP8 elements

* Scaled type convert for vectors of 16 FP8 elements

* Implementation of scaled conversion from F32 to F8

* Add tests for scaled conversions from FP32 to FP8

* Add documentation to the test functions

* Implementation of scaled conversion from F32x2 to F8x2

* Implementation of scaled conversion from F32x16 to F8x16

* Implementation of scaled conversion from F32x32 to F8x32

* Implementation of scaled conversion from F8x32 to F32x32

* Verified on the emulator

* MX FP GEMM - Example Template (#277)

Temporarily uses `DeviceGemmMultiD_ABScale_Xdl_CShuffle_V3` kernel and 128x128 scaling matrices.
Must be modified to use MX-native GEMM kernell with 16 or 32 component vectors per scale.

Verified on the emulator.

* Add vector support

* Add tests

* Add missing type aliases

* Fix test naming

* only build mx example for gfx950

* disable CK_USE_AMD_MFMA_GFX950 by default

* fic build for multiple archs

* fix typo

* fix typo

* Update unpack signature

* Fix merge

* Add size checks in pack function

* Add a flag

* Add conversions

* Fix build logic

* Update pack/unpack methods

* Remove unneeded AsType accessors

* Add docstrings

* Add a flag to config file

* Test the functionality of V_MFMA_F32_16X16X128_F8F6F4 and  V_MFMA_F32_32X32X64_F8F6F4 instructions. (#293)

* Introduced MFMA tests

* Verified f8f6f4 MFMA Instructions

* Move flag logic to scaled_type_convert header

* Use pointers instead of array indices

* Fix a typo

* Update tests and pack functions

* Fix gemm gemm on gfx950

* Fix clang format

* restore the default gput target lists

* fix the jenkinsfile

* add missing ifdef

---------

Co-authored-by: Jing Zhang <jizhan@amd.com>
Co-authored-by: aska-0096 <haocwang@amd.com>
Co-authored-by: Jun Liu <Liu.Jun@amd.com>
Co-authored-by: Andriy Roshchenko <andriy.roshchenko@amd.com>
Co-authored-by: Rostyslav Geyyer <rosty.geyyer@amd.com>
Co-authored-by: Rostyslav Geyyer <46627076+geyyer@users.noreply.github.com>
Co-authored-by: root <root@banff-cyxtera-s83-2.ctr.dcgpu>
Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>
Co-authored-by: jefyang1 <146495389+jefyang1@users.noreply.github.com>
Co-authored-by: jefyang1 <Jeffreyj.Yang@amd.com>

aska-0096 committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into update_cka8w8_uc

aska-0096 committed:
clang format

aska-0096 committed:
fix errors

aska-0096 committed:
fix errors

aska-0096 committed:
remove compile flags in example

aska-0096 committed:
fix error

spolifroni-amd reviewed (commented):
Doesn't look like there's anything for docs to review. Let me know if I'm wrong about that.

illsilin committed:
restore cron trigger (#1863)

zjing14 commented:
@aska-0096 It is great to see weight pre-shuffling optimization in CK GEMM. Looking forward to see some preliminary performance results.

aska-0096 committed:
recover enable-post-misched=0 for sanity issue

jakpiase committed:
add vectorloads on non-k dim for memory pipelines (#1856)

kylasa committed:
Support for dtypes (fp8, bf8, bf16 and fp16) for the ck_tile/03_gemm example. (#1845)

* Support bf16/fb8/bf8 datatypes for ck_tile/gemm

* remove commented out code.

* Addressing code review comments and enabling universal_gemm for all the supported data types.

* Merge conflict resolution.

* Solve the memory pipeline compilation error. Merge with the new change of CShuffle

* finish the feature, pass the tests

* Fix the pipeline and add the benchmark script for other data types

---------

Co-authored-by: ThomasNing <thomas.ning@amd.com>

aska-0096 committed:
revert blockwisegemm modification

aska-0096 committed:
revert blkgemm pipe v2 changes.

aosewski committed:
CK Tile - small fix to hotloop scheduler & KPack value. (#1867)

* Use SmemPack in HotLoop scheduler

* Additional debug print information

* Change KPack value.

Hardcode for now, as without AK1/BK1 there's no good way to determine
its value.

* Fix HotLoopScheduler MFMA instr parameters.

geyyer committed:
Add a host mx gemm reference kernel (#1864)

* Add mx gemm reference kernel

* Update copyright year

* Update mx gemm example

* Use element-wise ops in the reference gemm

danielsu-amd committed:
External CI: enable amd-develop branch trigger (#1859)

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

aska-0096 committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into update_cka8w8_uc

aska-0096 commented:
Hi @illsilin 
CI passed.

afagaj commented on example/65_gemm_multiply_multiply/CMakeLists.txt:
```diff
@@ -1,4 +1,7 @@
 add_example_executable(example_gemm_multiply_multiply_xdl_fp8 gemm_multiply_multiply_xdl_fp8.cpp)
+# target_compile_options(example_gemm_multiply_multiply_xdl_fp8 PRIVATE -mllvm -greedy-reverse-local-assignment=1 -save-temps=$PWD -Wno-gnu-line-marker)
```
Had you meant to remove this line?
```suggestion
```

afagaj commented on example/65_gemm_multiply_multiply/CMakeLists.txt:
```diff
@@ -1,4 +1,7 @@
+# target_compile_options(example_gemm_multiply_multiply_xdl_fp8 PRIVATE -mllvm -greedy-reverse-local-assignment=1 -save-temps=$PWD -Wno-gnu-line-marker)
 add_example_executable(example_gemm_multiply_multiply_xdl_fp8_ab_scale gemm_multiply_multiply_xdl_fp8_ab_scale.cpp)
+add_example_executable(example_gemm_multiply_multiply_xdl_fp8_bpreshuffle gemm_multiply_multiply_xdl_fp8_bpreshuffle.cpp)
+# target_compile_options(example_gemm_multiply_multiply_xdl_fp8_bpreshuffle PRIVATE -mllvm -greedy-reverse-local-assignment=1 -save-temps=$PWD -Wno-gnu-line-marker)
```
Had you meant to remove this line?
```suggestion
```

afagaj commented on example/65_gemm_multiply_multiply/gemm_multiply_multiply_xdl_fp8_bpreshuffle.cpp:
```diff
@@ -0,0 +1,397 @@
+
+static constexpr auto GemmSpec = ck::tensor_operation::device::GemmSpecialization::Default;
+
+// using DeviceOpInstance = ck::tensor_operation::device::DeviceGemmMultiD_Xdl_CShuffle_V3
```
Had you meant to remove this line?
```suggestion
```

afagaj commented on profiler/include/profiler/profile_gemm_multiply_multiply_impl.hpp:
```diff
@@ -188,7 +188,9 @@ bool profile_gemm_multiply_multiply_impl(int do_verification,
     {
-        std::vector<int> kbatch_list = {1, 2, 4, 8, 16, 19, 32, 38};
+        // Seems like when performance measurement has bug when spiltK is large
+        // std::vector<int> kbatch_list = {1, 2, 4, 8, 16, 19, 32, 38};
```
Had you meant to delete this?
```suggestion
```

afagaj commented on profiler/include/profiler/profile_gemm_multiply_multiply_impl.hpp:
```diff
@@ -249,6 +270,10 @@ bool profile_gemm_multiply_multiply_impl(int do_verification,
                 std::string op_name = op_ptr->GetTypeString();
 
+                // timer of develop branch should only apply to empty hipstream
+                // hipStream_t stream;
+                // hip_check_error(hipStreamCreate(&stream));
```
Was this code meant to be uncommented?

aska-0096 replied:
This code is a kind of explanatory note to whoever wondering why we use null hipStream in the ckProfiler kernel launch,

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

aska-0096 committed:
Apply suggestions from code review

Co-authored-by: John Afaganis <john.afaganis@amd.com>

aska-0096 committed:
Merge branch 'develop' into update_cka8w8_uc

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

aska-0096 committed:
hotfix for ckprofiler operator

aska-0096 committed:
Merge branch 'update_cka8w8_uc' of https://github.com/ROCm/composable_kernel into update_cka8w8_uc

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

aska-0096 committed:
Merge branch 'develop' into update_cka8w8_uc

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

ThomasNing committed:
Merge branch 'develop' into update_cka8w8_uc

ThomasNing committed:
add the 16x16 mfma instances

ThomasNing commented:
Already add the 16x16 instances to the CK Profiler for pre-shuffle.

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

ThomasNing committed:
Add more instances for 16x16 instructions

aosewski reviewed (dismissed):
Are the changes to `rmsnorm2d_rdquant` relevant to this PR or is it a bad merge artifact?

ThomasNing commented:
@aosewski It is a feature adding in this PR to add the quant data type for rmsnorm2d_rdquant ck_tile operator.

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

ThomasNing committed:
Merge branch 'develop' into update_cka8w8_uc
